### PR TITLE
Support spreading and destructuring of iterators in ES6

### PR DESCRIPTION
**Pull request by JsonFreeman**
This adds support for a few things:
* Spreading an iterator into an array
```ts
var array = [0, 1, ...new NumberIterator];
```
* Spreading an iterator into a call
```ts
function foo(...nums: number[]) { }
foo(...new NumberIterator);
```
* Destructuring an iterator into an array
```ts
var [a, b, ...c] = new NumberIterator; // a and b have type number, c has type number[]
var d: number, e: number, f: number[];
[d, e, ...f] = new NumberIterator;
function foo([p, q, ...r]: Iterable<number>) {}
foo(new NumberIterator);
```

I will also add that tuples are relatively unaffected. If you declare something to be a tuple, you can still only assign a tuple to it. And if you have an array binding pattern with no rest element, its implied type is a tuple. So the following does give you an error:
```ts
function foo([p, q]) {}
foo(new NumberIterator);
```

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2153,7 +2158,34 @@ module ts {
...
+                let unionOfElements = getUnionType(elementTypes);
+                if (languageVersion >= ScriptTarget.ES6) {
+                    // If the user has something like:
+                    //
+                    //     function fun(...[a, ...b]) { }
+                    //
+                    // Normally, in ES6, the implied type of an array binding pattern with a rest element is
+                    // an iterable. However, there is a requirement in our grammar checker that all rest
+                    // parameters be array types. To satisfy this, we have an exception to the rule that
+                    // says the type of an array binding pattern with a rest element is an array type
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: That's an odd requirement given that the grammar only says this:

```
RestParameter:
    ...   Identifier   TypeAnnotation_opt
```

I don't see why, at grammar check time, we'd care if the parameter was an array type or not.

Comment (JsonFreeman) [inline src/compiler/checker.ts]: It's the type system, not the grammar checker. I will fix the comment

**Comment by JsonFreeman** [comment]
To see a design proposal, see https://github.com/Microsoft/TypeScript/issues/1900#issuecomment-76583563

**Comment by yuit** [comment]
forgot to approve :+1:

**Pull request merged**